### PR TITLE
smartcontract: update nns wrapper

### LINKED ISSUE #2732: NNS wrapper update
We have an official NNS deployed to mainnet now which means that it has a stable API and we can provide a wrapper for it. One thing I don't like about it is that it's a non-native contract and keeping this wrapper in the `pkg/rpcclient` seems to be inappropriate. At the same time it'll be build on top of `rpcclient` and we'll need it at least in the wallet code. Maybe some new top-level directory is needed for it, we can have more of these wrappers. Or a separate repository, but that'd make a circular dependency which is not a good thing.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Added missing function from nns smart-contract manifest.

Closes #2732.

### CODE REVIEW

File: pkg/rpcclient/nns/contract.go [hunk 1]
```diff
@@ -44,6 +71,21 @@ func NewReader(invoker Invoker, hash util.Uint160) *ContractReader {
 	return &ContractReader{*nep11.NewNonDivisibleReader(invoker, hash), invoker, hash}
 }
 
+// Hash stores the hash of the native NEOToken contract.
+var Hash = state.CreateNativeContractHash(nativenames.Neo)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:75]: Please, be more careful. It's NNS.

File: pkg/rpcclient/nns/contract.go [hunk 2]
```diff
@@ -31,6 +36,28 @@ type ContractReader struct {
 	hash    util.Uint160
 }
 
+// Actor is used by Contract to call state-changing methods.
+type Actor interface {
+	Invoker
+
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:42]: Useless empty line.

File: pkg/rpcclient/nns/contract.go [hunk 3]
```diff
@@ -31,6 +36,28 @@ type ContractReader struct {
...
+	MakeCall(contract util.Uint160, method string, params ...any) (*transaction.Transaction, error)
+	MakeRun(script []byte) (*transaction.Transaction, error)
+	MakeUnsignedCall(contract util.Uint160, method string, attrs []transaction.Attribute, params ...any) (*transaction.Transaction, error)
+	MakeUnsignedRun(script []byte, attrs []transaction.Attribute) (*transaction.Transaction, error)
+	SendCall(contract util.Uint160, method string, params ...any) (util.Uint256, uint32, error)
+	SendRun(script []byte) (util.Uint256, uint32, error)
+}
+
+// Contract implements all contract methods.
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:53]: Make comment similar to Neo: `// Contract provides full NeoNameService interface, both safe and state-changing methods.`

File: pkg/rpcclient/nns/contract.go [hunk 4]
```diff
@@ -31,6 +36,28 @@ type ContractReader struct {
...
+type Contract struct {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:54]: Need to adjust documentation to `nns` package.

File: pkg/rpcclient/nns/contract.go [hunk 5]
```diff
@@ -31,6 +36,28 @@ type ContractReader struct {
...
+	ContractReader
+	nep11.BaseWriter
+	actor Actor
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:57]: Need blank line to distinguish interfaces from unexported fields.

File: pkg/rpcclient/nns/contract.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go:75]: BTW, we need a dynamic hash for this wrapper. Have you tried to run `generate-rpcwrapper` command for NNS manifest deployed on mainnet? Run it without `--hash` flag. Take a look at the output. It may be taken as base.

File: pkg/rpcclient/nns/contract.go [hunk 6]
```diff
@@ -120,3 +162,231 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) SetPrice(priceList []any) (util.Uint256, uint32, error) {
+	return c.actor.SendCall(c.hash, "setPrice", priceList)
+}
+
+// SetPriceTransaction creates a transaction invoking `setPrice` method of the contract.
+// This transaction is signed, but not sent to the network, instead it's
+// returned to the caller.
+func (c *Contract) SetPriceTransaction(priceList []any) (*transaction.Transaction, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:234]: Why `[]any` is used? We need this wrapper to be as precise as possible.

File: pkg/rpcclient/nns/contract.go [hunk 7]
```diff
@@ -120,3 +162,231 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+	return unwrap.SessionIterator(c.invoker.Call(c.hash, "roots"))
+}
+
+// RootsExpanded is similar to Roots (uses the same contract
+// method), but can be useful if the server used doesn't support sessions and
+// doesn't expand iterators. It creates a script that will get the specified
+// number of result items from the iterator right in the VM and return them to
+// you. It's only limited by VM stack and GAS available for RPC invocations.
+func (c *ContractReader) RootsExpanded(_numOfIteratorItems int) ([]stackitem.Item, error) {
+	return unwrap.Array(c.invoker.CallAndExpandIterator(c.hash, "roots", _numOfIteratorItems))
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:177]: We need return value to be some meaningful Go type, not an array of stackitems.

File: pkg/rpcclient/nns/contract.go [hunk 8]
```diff
@@ -120,3 +162,231 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) SetAdminUnsigned(name string, admin util.Uint160) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "setAdmin", nil, name, admin)
+}
+
+// SetRecord creates a transaction invoking `setRecord` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) SetRecord(name string, typev *big.Int, data string) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:353]: There's a set of record types defined via `RecordType` type. We need to use it where possible.

File: pkg/rpcclient/nns/contract.go (hunk 7 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go:177]: Take a look at the way how `GetAllRecordsExpanded` is handled.

File: pkg/rpcclient/nns/contract_test.go [hunk 9]
```diff
@@ -13,7 +14,13 @@ import (
 
 type testAct struct {
 	err error
+	ser error
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:17]: What is this?

File: pkg/rpcclient/nns/contract_test.go [hunk 10]
```diff
@@ -13,7 +14,13 @@ import (
...
 	res *result.Invoke
+	rre *result.Invoke
+	rer error
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:20]: Meaningless names, need to rename or add a comment.

File: pkg/rpcclient/nns/contract.go (hunk 4 above)
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:54]: Have you compared existing methods with mainnet NNS? Our goal is to have all methods compatible with it. If needed, then existing methods may be changed.

**Review by AnnaShaleva — COMMENTED** [review]
Add reference to initial issue to commit/PR description.

File: pkg/rpcclient/nns/contract.go (hunk 4 above)
Comment (AliceInHunterland) [reply pkg/rpcclient/nns/contract.go:54]: yes, i hope i didn't miss any of them

Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go:51]: Not resolved. Search for `Package nns provide some RPC wrappers for the non-native NNS contract.` and adjust this documentation.

File: pkg/rpcclient/nns/contract.go [hunk 11]
```diff
@@ -44,6 +69,21 @@ func NewReader(invoker Invoker, hash util.Uint160) *ContractReader {
 	return &ContractReader{*nep11.NewNonDivisibleReader(invoker, hash), invoker, hash}
 }
 
+// Hash contains contract hash.
+var Hash = util.Uint160{0xde, 0x46, 0x5f, 0x5d, 0x50, 0x57, 0xcf, 0x33, 0x28, 0x47, 0x94, 0xc5, 0xcf, 0xc2, 0xc, 0x69, 0x37, 0x1c, 0xac, 0x50}
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:73]: We need a dynamic hash. Run `generate-rpcwrapper` without `--hash` flag and merge s the result with an existing code. There shouldn't be `Hash` field.

File: pkg/rpcclient/nns/contract.go [hunk 12]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// method), but can be useful if the server used doesn't support sessions and
+// doesn't expand iterators. It creates a script that will get the specified
+// number of result items from the iterator right in the VM and return them to
+// you. It's only limited by VM stack and GAS available for RPC invocations.
+func (c *ContractReader) RootsExpanded(_numOfIteratorItems int) ([]RecordState, error) {
+	arr, err := unwrap.Array(c.invoker.CallAndExpandIterator(c.hash, "roots", _numOfIteratorItems))
+	if err != nil {
+		return nil, err
+	}
+	return itemsToRecords(arr)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:179]: NACK. Have you ever tried to invoke this method on real NNS in testnet? Try, and it will fail, because the returned value of `roots` is array of strings.

File: pkg/rpcclient/nns/contract.go [hunk 13]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
 	}
 	return res, nil
 }
+
+// Roots invokes `roots` method of contract.
+func (c *ContractReader) Roots() (uuid.UUID, result.Iterator, error) {
+	return unwrap.SessionIterator(c.invoker.Call(c.hash, "roots"))
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:166]: We need a special iterator type for it. Like `RecordIterator` but it should iterate over records (strings).

File: pkg/rpcclient/nns/contract.go [hunk 14]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+}
+
+// RootsExpanded is similar to Roots (uses the same contract
+// method), but can be useful if the server used doesn't support sessions and
+// doesn't expand iterators. It creates a script that will get the specified
+// number of result items from the iterator right in the VM and return them to
+// you. It's only limited by VM stack and GAS available for RPC invocations.
+func (c *ContractReader) RootsExpanded(_numOfIteratorItems int) ([]RecordState, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:174]: Missing `GetRecord` wrapper.

File: pkg/rpcclient/nns/contract.go [hunk 15]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) AddRootUnsigned(root string) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "addRoot", nil, root)
+}
+
+// SetPrice creates a transaction invoking `setPrice` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) SetPrice(priceList []any) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:229]: Still `[]any`, I'd expect you to change it *everywhere* where appropriate.

File: pkg/rpcclient/nns/contract.go [hunk 16]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) SetAdminUnsigned(name string, admin util.Uint160) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "setAdmin", nil, name, admin)
+}
+
+// SetRecord creates a transaction invoking `setRecord` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) SetRecord(name string, typev RecordType, data string) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:355]: Have you tried to invoke it? Does it work with `RecordType`?

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/rpcclient/nns/contract.go [hunk 17]
```diff
@@ -120,3 +160,235 @@ func itemsToRecords(arr []stackitem.Item) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) RenewUnsigned(name string) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "renew", nil, name)
+}
+
+// Renew2 creates a transaction invoking `renew` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) Renew2(name string, years *big.Int) (util.Uint256, uint32, error) {
```
[the remark this answers is not in the document]
Comment (AliceInHunterland) [reply pkg/rpcclient/nns/contract.go:311]: linter wasnt happy about it

[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go:311]: Ref. https://github.com/nspcc-dev/neo-go/issues/3296.

File: pkg/rpcclient/nns/contract_test.go [hunk 18]
```diff
@@ -195,3 +222,38 @@ func TestGetAllRecordsExpanded(t *testing.T) {
 		Data: "cool",
 	}, vals[0])
 }
+
+func TestRoots(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AliceInHunterland) [reply pkg/rpcclient/nns/contract_test.go:227]: and accidentally find out that TestClient_IteratorSessions - fails. but it seems it doesn't connect with this PR.

**Comment by AnnaShaleva** [comment]
Linter is failing due to underscores in func names, so let's resolve this issue in #3296, and for now let's have underscore-free names.

File: pkg/rpcclient/nns/contract.go [hunk 19]
```diff
@@ -88,28 +151,417 @@ func (r *RecordIterator) Next(num int) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) SetAdminUnsigned(name string, admin util.Uint160) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "setAdmin", nil, name, admin)
+}
+
+// SetRecord creates a transaction invoking `setRecord` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) SetRecord(name string, typev *big.Int, data string) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:358]: Replace `typev` type.

File: pkg/rpcclient/nns/contract.go [hunk 20]
```diff
@@ -88,28 +151,417 @@ func (r *RecordIterator) Next(num int) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) SetRecordUnsigned(name string, typev *big.Int, data string) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "setRecord", nil, name, typev, data)
+}
+
+// DeleteRecord creates a transaction invoking `deleteRecord` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) DeleteRecord(name string, typev *big.Int) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:380]: Replace type of typev.

File: pkg/rpcclient/nns/contract_test.go [hunk 21]
```diff
@@ -29,12 +34,35 @@ func (t *testAct) TraverseIterator(sessionID uuid.UUID, iterator *result.Iterato
...
+	return t.txh, t.vub, t.err
+}
+
 func TestSimpleGetters(t *testing.T) {
 	ta := &testAct{}
 	nns := NewReader(ta, util.Uint160{1, 2, 3})
 
 	ta.err = errors.New("")
-	_, err := nns.GetPrice()
+	_, err := nns.GetPrice(new(big.Int).SetInt64(int64(A)))
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:65]: The argument of GetPrise is name length, not the record type.

File: pkg/rpcclient/nns/contract_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract_test.go:227]: > and accidentally find out that TestClient_IteratorSessions - fails. but it seems it doesn't connect with this PR.

Issue, please. It passes in our MacOS testing job.

File: pkg/rpcclient/nns/contract_test.go [hunk 22]
```diff
@@ -195,3 +223,135 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+	require.Error(t, err)
+
+	// Test success case
+	ta.err = nil
+	ta.txh = util.Uint256{0x0A, 0x0B, 0x0C}
+	txh, vub, err := nns.SetPrice(priceList)
+	require.NoError(t, err)
+	require.Equal(t, ta.txh, txh)
+	require.Equal(t, ta.vub, vub)
+}
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:318]: OK, I see that test is here. But it doesn't call `NewParameterFromValue`. `SetPrice` will fail on real invocation, we need to adjust the method itself to provide `[]any` to `SendCall` instead of `[]*big.Int`.

Comment (AliceInHunterland) [reply pkg/rpcclient/nns/contract_test.go]: so SetPrice should be: SetPrice(priceList []any) ?

File: pkg/rpcclient/nns/contract_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (AliceInHunterland) [reply pkg/rpcclient/nns/contract_test.go:228]: #3300

File: pkg/rpcclient/nns/contract.go [hunk 23]
```diff
@@ -1,68 +1,127 @@
...
 
-// IsAvailable checks whether the domain given is available for registration.
+// Roots invokes `roots` method of contract.
+func (c *ContractReader) Roots() (*RootsIterator, error) {
+	sess, iter, err := unwrap.SessionIterator(c.invoker.Call(c.hash, "roots"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &RootsIterator{
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:88]: BTW, `RootIterator`? To match the `RecordIterator`.

File: pkg/rpcclient/nns/contract.go (hunk 15 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go:227]: Not resolved.

File: pkg/rpcclient/nns/contract.go [hunk 24]
```diff
@@ -88,35 +147,430 @@ func (r *RecordIterator) Next(num int) ([]RecordState, error) {
...
+// Any fields of it that do not affect fees can be changed (ValidUntilBlock,
+// Nonce), fee values (NetworkFee, SystemFee) can be increased as well.
+func (c *Contract) RenewUnsigned(name string) (*transaction.Transaction, error) {
+	return c.actor.MakeUnsignedCall(c.hash, "renew", nil, name)
+}
+
+// Renew2 creates a transaction invoking `renew` method of the contract.
+// This transaction is signed and immediately sent to the network.
+// The values returned are its hash, ValidUntilBlock value and error if any.
+func (c *Contract) Renew2(name string, years int64) (util.Uint256, uint32, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:309]: Should update related methods parameters correspondingly (from bigint to int64).

File: pkg/rpcclient/nns/iterators.go [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+
+// RecordIterator is used for iterating over GetAllRecords results.
+type RecordIterator struct {
+	client   Invoker
+	session  uuid.UUID
+	iterator result.Iterator
+}
+
+// RootsIterator is used for iterating over Roots results.
+type RootsIterator struct {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/iterators.go:20]: Move `Next` method defined on these iterators to the same file.

File: pkg/rpcclient/nns/contract_test.go [hunk 26]
```diff
@@ -195,3 +224,185 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+
+func TestIntTransactions(t *testing.T) {
+	ta := new(testAct)
+	nns := New(ta, util.Uint160{1, 2, 3})
+
+	for _, fun := range []func(nef []byte, manifest string) (*transaction.Transaction, error){
+		nns.UpdateTransaction,
+		nns.UpdateUnsigned,
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:396]: Test name is `IntTransactions`, and the content is about update.

Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract_test.go:396]: And you already have the test for `update`, use it for `UpdateTransaction` and `UpdateUnsigned` tests.

**Comment by AnnaShaleva** [comment]
Need to be rebased onto master.

File: pkg/rpcclient/nns/contract_test.go (hunk 22 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract_test.go]: It should be `SetPrice(priceList []int64)`, and inside you should convert `[]int64` to `[]any`, I guess.

File: pkg/rpcclient/nns/contract.go [hunk 27]
```diff
@@ -1,68 +1,127 @@
...
 func (c *ContractReader) IsAvailable(name string) (bool, error) {
 	return unwrap.Bool(c.invoker.Call(c.hash, "isAvailable", name))
 }
 
-// Resolve resolves the given record type for the given domain (with no more
-// than three redirects).
-func (c *ContractReader) Resolve(name string, typ RecordType) (string, error) {
-	return unwrap.UTF8String(c.invoker.Call(c.hash, "resolve", name, int64(typ)))
+// GetRecord invokes `getRecord` method of contract.
+func (c *ContractReader) GetRecord(name string, typev *RecordType) (string, error) {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:120]: Why record type passed as a pointer?

File: pkg/rpcclient/nns/contract.go [hunk 28]
```diff
@@ -88,35 +147,430 @@ func (r *RecordIterator) Next(num int) ([]RecordState, error) {
...
+	if err != nil {
+		return fmt.Errorf("field NewExpiration: %w", err)
+	}
+
+	return nil
+}
+
+// Terminate closes the iterator session used by RecordIterator (if it's
+// session-based).
+func (r *RecordIterator) Terminate() error {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:562]: These two should also be moved to `iterators.go`.

**Review by AnnaShaleva — COMMENTED** [review]
And we need to think about the destination place for NNS wrapper because it's not native contract anymore.

File: pkg/rpcclient/nns/contract.go [hunk 29]
```diff
@@ -1,68 +1,131 @@
...
+	arr, err := unwrap.Array(c.invoker.CallAndExpandIterator(c.hash, "roots", _numOfIteratorItems))
+	if err != nil {
+		return nil, err
+	}
+
+	return itemsToRoots(arr), nil
+}
+
+// GetPrice invokes `getPrice` method of contract.
+func (c *ContractReader) GetPrice(length *big.Int) (*big.Int, error) {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract.go]: Not resolved.

File: pkg/rpcclient/nns/contract.go [hunk 30]
```diff
@@ -1,68 +1,127 @@
-/*
-Package nns provide some RPC wrappers for the non-native NNS contract.
-
-It's not yet a complete interface because there are different NNS versions
-available, yet it provides the most widely used ones that were available from
-the old RPC client API.
-*/
+// Package nns provide RPC wrappers for the non-native NNS contract.
+// This is Neo N3 NNS contract wrapper, the source code of the contract can be found here:
+// https://github.com/neo-project/non-native-contracts
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:3]: Need a specific link to NNS and git revision in the link.

File: pkg/rpcclient/nns/contract.go [hunk 31]
```diff
@@ -1,68 +1,127 @@
...
 	"github.com/nspcc-dev/neo-go/pkg/rpcclient/nep11"
 	"github.com/nspcc-dev/neo-go/pkg/rpcclient/unwrap"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
 	"github.com/nspcc-dev/neo-go/pkg/util"
 	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
 )
 
-// Invoker is used by ContractReader to call various methods.
+// SetAdminEvent represents "SetAdmin" event emitted by the contract.
+type SetAdminEvent struct {
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract.go:22]: Please, add `MaxNameLenght` constant with 255 value and appropriate comment.

File: pkg/rpcclient/nns/contract_test.go [hunk 32]
```diff
@@ -195,3 +224,180 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+	require.Equal(t, ta.vub, vub)
+}
+
+func TestSetPrice(t *testing.T) {
+	ta := &testAct{}
+	nns := New(ta, util.Uint160{1, 2, 3})
+
+	priceList := []int64{100, 200}
+	_, err := smartcontract.NewParameterFromValue(priceList)
+	require.NoError(t, err)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:352]: You'll receive error here because NewParameterFromValue doesn't handle  `[]int64`. And these two lines are not related to the test, remove them.

File: pkg/rpcclient/nns/contract_test.go [hunk 33]
```diff
@@ -195,3 +224,180 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+	ta.err = errors.New("test error")
+	_, _, err = nns.SetPrice(priceList)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:354]: Test for `SetPriceTransaction` and `SetPriceUnsigned`?

File: pkg/rpcclient/nns/contract_test.go [hunk 34]
```diff
@@ -195,3 +224,180 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+	owner := util.Uint160{0x0D, 0x0E, 0x0F}
+
+	ta.err = errors.New("test error")
+	_, _, err := nns.Register(name, owner)
+	require.Error(t, err)
+
+	// Test success case
+	ta.err = nil
+	ta.txh = util.Uint256{0x10, 0x11, 0x12}
+	txh, vub, err := nns.Register(name, owner)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:380]: Transaction/Unsigned test.

File: pkg/rpcclient/nns/contract_test.go [hunk 35]
```diff
@@ -195,3 +224,180 @@ func TestGetAllRecordsExpanded(t *testing.T) {
...
+}
+
+func TestRenew(t *testing.T) {
+	ta := &testAct{}
+	nns := New(ta, util.Uint160{1, 2, 3})
+
+	name := "example.neo"
+
+	ta.err = errors.New("test error")
+	_, _, err := nns.Renew(name)
```
Comment (AnnaShaleva) [inline pkg/rpcclient/nns/contract_test.go:393]: Transaction/Unsigned test.

File: pkg/rpcclient/nns/contract_test.go (hunk 33 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nns/contract_test.go:354]: Take a look at `TestRegisterCandidate` of native Neo wrapper: there are error cases for three setters and good cases for the same three setters.

**Review by AnnaShaleva — COMMENTED** [review]
s/`Refs #2732`/`Close #2732.` in the commit message, because this single commit closes the whole issue. Also, in the same commit we need to move this wrapper to another directory.

@roman-khimov, I've consider the `pkg/rpcwrappers` destination directory for this package, because currently we have `rpcclient` upper-level directory with `rpc` prefix, so it looks similar. I also considered something like `pkg/nonnative-wrappers` if one day we're going to have simple wrappers for contracts (but it's only applicable for go contracts, so hardly ever we'll add something). Maybe you have something in mind? Or maybe @AliceInHunterland?

**Comment by roman-khimov** [comment]
`rpcwrappers` seem to be redundant here. Ideally, external contracts would have wrappers in their respective repositories (like neofs-contract), in this case that'd be https://github.com/neo-project/non-native-contracts. This is perfectly fine for most of contracts. It can be a separate repo as well if it's not accepted into the upstream one and that's also OK in most cases. The problem is that we're going to use this wrapper in our own CLI, so suddenly we have a circular dependency. Which is not the end of the world (hi, https://github.com/nspcc-dev/dbft/issues/2), but still an inconvenience, especially if we're considering updates.

Many other rpcclient subpackages are wrappers in their essence as well.

**Comment by AnnaShaleva** [comment]
> external contracts would have wrappers in their respective repositories

Not the NNS case, hardly ever we'll merge these changes in the https://github.com/neo-project/non-native-contracts.

> It can be a separate repo as well

I saw this comment in the original issue (https://github.com/nspcc-dev/neo-go/issues/2732#issue-1400791400), but I don't want to introduce this circular dependency, because it's irritating to update it. I don't want to add more updating problems, we already have interop deps updating scripts, it's enough I think.

> Many other rpcclient subpackages are wrappers in their essence as well.

Do you suggest to keep it as is in the `pkg/rpcclient/nns`?

**Comment by roman-khimov** [comment]
Yep, `pkg/rpclient/nns` seems to be OK given all the constraints, `rpcwrappers` just don't improve much.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland, please, move the wrapper back to `rpcclient/nns`.

**Review by AnnaShaleva — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]
